### DISCUSSION TITLE
Problem with CORS policy in React front end with FastAPI on the back trying to use Microsoft Single Sign-on

### DISCUSSION THREAD

**Discussion by fbuenohernandez** [fastapi/fastapi#12203] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import React, { useState } from 'react';
import './App.css';

function App() {
  
    const handleSubmit = (event) => {
    event.preventDefault(); // Evita que a página seja recarregada ao enviar o formulário
    
        // URL sem os parâmetros GET
    const url = 'http://localhost:5050/login';

    fetch(url,  
      {method: 'GET',
}).then((response) => {
        if (!response.ok) {
          throw new Error('Erro na solicitação');
        }

        window.location.href = response.json()
        return response.json();
      })
      .then((data) => {
        console.log('Resposta do servidor:', data);
        })
      .catch((error) => {
        console.error('Erro:', error);
      });
  };

  return (
    <div className="login-container">
      <h2>Login</h2>
      <form onSubmit={handleSubmit}>
                
        <button type="submit" className="login-button">Logar</button>
      </form>
    </div>
  );
}

--------------------------------------------------------------------------------------------------

from authlib.integrations.starlette_client import OAuth
import uvicorn
import logging
from fastapi import FastAPI, Request
from starlette.middleware.sessions import SessionMiddleware
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

logger = logging.getLogger('uvicorn.error')

app.add_middleware(
    SessionMiddleware,
    secret_key="?",  # You should use a strong, random key
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

oauth = OAuth()
oauth.register(
    name='azure',
    client_id='SEU_CLIENT_ID',
    client_secret='SEU_CLIENT_SECRET',
    authorize_url='https://login.microsoftonline.com/SEU_TENANT_ID/oauth2/v2.0/authorize',
    token_url='https://login.microsoftonline.com/SEU_TENANT_ID/oauth2/v2.0/token',
    client_kwargs={'scope': 'openid email profile User.Read'}
)

@app.api_route('/login')
async def login(request: Request):
    redirect_uri = request.url_for('auth')
    return await oauth.azure.authorize_redirect(request, redirect_uri)

@app.get("/callback")
async def auth(request: Request):
    token = await oauth.azure.authorize_access_token(request)
    user = await oauth.azure.parse_id_token(request, token)
    return {"token": token, "user_info": user}

@app.get("/")
def hello_world_root():
    return {"Hello": "World"}

if __name__ == '__main__':
    uvicorn.run(app, port=5050)
```

### Description

I want to be redirect when clicking the button on the React front. 

What I get:

Access to fetch at 'http://localhost:5050/login' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

The root route works just fine, if I access the ip:port of my backend I get redirected to the SSO perfectly, so the Azure credentials are configured right.

You probably noticed I have two Middlewares on the code, I've tried both separately and together.

If I don't use starlette's I get: "session" in self.scope AssertionError: SessionMiddleware must be installed to access request.session

The CORSmiddleware is as useful as a solar-powered flashlight.

I would appreciate some help here.

Regards

### Operating System

Windows

### FastAPI Version

0.114.1

### Pydantic Version

None

### Python Version

3.12.4

**Comment by sinisaos** [comment]
@fbuenohernandez If you set `allow_credentials` to `True`, then you cannot use `*` (from security point of view, wildcard is too permissive and would defeat use of credentials). You will need to provide the exact url in the origins. Something like this
```python
origins = [
    "http://localhost:3000", # your frontend url
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```
Can you try that and see if it works? Sorry if I missed your point.

**Reply by fbuenohernandez** [reply]
I appreciate your reply @sinisaos, I'm open to any suggestions.

The error message changed, I see this as a win:

Access to fetch at 'https://login.microsoftonline.com/blablabla/oauth2/v2.0/authorize?response_type=code&client_id=blablabla&redirect_uri=http%3A%2F%2Flocalhost%3A5050%2Fauth&scope=openid+email+profile+User.Read&state=blablabla&nonce=blablabla' (redirected from 'http://localhost:5050/login') from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Reply by sinisaos** [reply, marked as answer]
@fbuenohernandez Your question (or a very similar question) seems to have been answered on [Stackoverflow](https://stackoverflow.com/questions/78984218/problem-with-cors-policy-in-react-front-end-with-fastapi-on-the-back-trying-to-u) with working code examples. Try that suggestion and see if it works. From that answer, everything seems to work. Good luck.

**Reply by fbuenohernandez** [reply]
It's me there :).

Regards
